### PR TITLE
INCOMING_MODULE_JS_API

### PR DESCRIPTION
**Pull request by kripken**
This new setting lets the user define which Module arguments will be provided at runtime. By default we assume, as always, that the user may provide any such API. With this new setting, the compiler can be told at compile time that only some of those values may be arriving. That allows it to not emit code for all the options that are not.

This starts to optimize using that option, but only does a little work so far. Specifically, `Module['arguments'], Module['thisProgram'], Module['quit']` are all optimized, that is, they are no longer placed on Module, and they are not read from Module unless the user expects them to be provided at runtime. (Again, that is the default, so no existing code should break.)

On hello world this saves 2.5%, but again, this option lets us optimize a lot more later. In principle, I think this option is key to letting us remove the bulk of unnecessary code from our minimal JS, in fact

### CODE REVIEW

File: src/settings.js [hunk 1]
```diff
@@ -578,6 +578,43 @@ var EXPORTED_RUNTIME_METHODS = [];
...
+// optimize.
+//
+// Setting this list to [], or at least a short and concise set of names you
+// actually use, can be very useful for reducing code size. By default the
+// list contains all the possible APIs.
+//
+// FIXME: should this just be  0  if we want everything?
+var INCOMING_MODULE_JS_API = [
+  'ENVIRONMENT', 'GL_MAX_TEXTURE_IMAGE_UNITS', 'SDL_canPlayWithWebAudio',
+  'SDL_numSimultaneouslyQueuedBuffers', 'TOTAL_MEMORYwasmMemory', 'arguments',
```
Comment (jgravelle-google) [inline src/settings.js]: s/'TOTAL_MEMORYwasmMemory'/'TOTAL_MEMORY', 'wasmMemory'/

Comment (kripken) [reply src/settings.js]: Thanks! Heh, this wasn't noticed because those aren't optimized yet, but eventually they should be...

**Comment by kripken** [comment]
Yeah, I somewhat agree with the concern about extra ifdefs. I think they aren't *that* bad, and the number of such ifdefs is at most the number of Module input values. Over time we may want to decrease those values, removing old APIs no one is using, etc.

I don't think there is an alternative to an approach like this, if we want maximal code size shrinking. Note that this is orthogonal to `MINIMAL_RUNTIME`, which removes a bunch of these by simply not having them, but it will still want some ifdefs for the few Module inputs it does care about.

I think more refactoring of our JS code can also improve the readability of these ifdefs. Smaller and more focused files, with more whitespace and comments, could go a long way. I'll try to do more of that as I add more optimizations.

**Comment by jgravelle-google** [comment]
I think there's hypothetical ways to do better, by say having a separate module per feature, that is included on-demand. In theory some of which are include-able without needing to use Emscripten directly! But that's very much an Emscripten2.0 type of thing, with no concrete demand and much more work needed to make useable.

So short of doing that: yeah I think this is as good as we're going to get it.

**Pull request merged**